docs/contributing-to-airbyte/building-new-connector/best-practices-for-connector-development.md

```diff
@@ -0,0 +1,53 @@
+2. **Fail fast.** A user should not be able to configure something that will not work. 
+3. **Fail actionably.** If a failure is actionable by the user, clearly let them know what they can do. Otherwise, make it very easy for them to give us necessary debugging information \(logs etc.\)
+
+From these principles we extrapolate the following goals for connectors, in descending priority order: 
+
+1. **Correct user input should result in a successful sync.** If there is an issue, it should be extremely easy for the user to see and report. 
+2. **Issues arising from bad user input should print an actionable error message.** "Invalid credentials" is not an actionable message. "Please verify your username/password is correct" is better. 
+3. **Wherever possible, a connector should support incremental sync.** This prevents excessive load on the underlying data source.  ****
+4. **When running a sync, a connector should communicate its status frequently to provide clear feedback that it is working.** Output a log message at least every 5 minutes. 
+5. **A connector should allow reading or writing as many entities as is feasible.** Supporting syncing all entities from an API is preferred to only supporting a small subset which would satisfy narrow use cases. Similarly, a database should support as many data types as is feasible. 
+
+Note that in the above list, the _least_ important is the number of features it has \(e.g: whether an API connector supports all entities in the API\). The most important thing is that for its declared features, it is reliable and usable. The only exception are “minimum viability” features e.g: for some sources, it’s not feasible to pull data without incremental due to rate limiting issues. In this case, those are considered usability issues.
+
+### Quality certification checklist 
+
+When reviewing connectors, we'll use the following "checklist" to verify whether the connector is considered "Airbyte certified" or closer to beta or alpha: 
+
+#### Integration Testing
+
+**As much as possible, prove functionality via testing**. This means slightly different things depending on the type of connector: 
+
+* **All connectors** must test all the sync modes they support during integration tests
+* **Database connectors** should test that they can replicate **all** supported data types in both `read` and `discover` operations
+* **API connectors** should validate records that every stream outputs data
+  * If this causes rate limiting problems, there should be a periodic CI build which tests this on a less frequent cadence to avoid rate limiting
+
+**Thoroughly test edge cases.** While Airbyte provides a [Standard Test Suite](testing-connectors.md#airbytes-standard-tests) that all connectors must pass, it's not possible for the standard test suite to cover all edge cases. When in doubt about whether the standard tests provide sufficient evidence of functionality, write a custom test case for your connector. 
```

Comment:
we will need a tutorial to write a custom test case.

Reply:
created issue